apiextensions: validation for customresources

nikhita opened the pull request:
- [x] Add types for validation of CustomResources
- [x] Fix conversion-gen: #49747
- [x] Fix defaulter-gen: kubernetes/gengo#61
- [x] Convert to OpenAPI types
- [x] Validate CR using go-openapi
- [x] Validate CRD Schema
- [x] Add integration tests
- [x] Fix round trip tests: #51204 
- [x] Add custom fuzzer functions
- [x] Add custom conversion functions
- [x] Fix data race while updating CRD: #50098 
- [x] Add feature gate for CustomResourceValidation
- [x] Fix protobuf generation

Proposal: https://github.com/kubernetes/community/pull/708
Additional discussion: https://github.com/kubernetes/kubernetes/issues/49879, https://github.com/kubernetes/kubernetes/pull/50625

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Add validation for CustomResources via JSON Schema.
```

/cc @sttts @deads2k

luxas commented:
@nikhita Please use two commits for this PR. One for your code and one for automatically generated code

@k8s-bot ok to test 

:smile:

nikhita committed:
Add types for validation of CustomResources

Remove protobuf generation because of the interface type

Add custom fuzzer funcs

Add custom marshalling

Add custom conversion functions

move jsonschema types to separate file

nikhita committed:
Add generated code

update generated proto

nikhita committed:
Validate CustomResource

* convert our types to openAPI types
* update strategy to include crd
* use strategy to validate customresource
* add helper funcs
* Fix conversion of empty ref field
* add validation for forbidden fields
* add defaulting for schema field
* Validate CRD Schema

nikhita committed:
Add integration tests

Update test schema

Add polling for TestCRValidationOnCRDUpdate

Add tests for forbidden fields

Enable featureGate for CustomResourceValidation

nikhita commented:
This will not pass tests until https://github.com/kubernetes/gengo/pull/61 is merged.

sttts commented on cmd/libs/go2idl/conversion-gen/generators/conversion.go:
```diff
@@ -338,36 +338,42 @@ func (e equalMemoryTypes) Equal(a, b *types.Type) bool {
 		return true
 	case in.Kind == out.Kind:
+		// if the type exists already, return early to avoid recursion
+		if existingTypes[in] {
```
also here: alreadyVisitedTypes

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -26,9 +30,10 @@ type CustomResourceDefinitionSpec struct {
 	// Scope indicates whether this resource is cluster or namespace scoped.  Default is namespaced
 	Scope ResourceScope
+	// Validation describes the validation methods for CustomResources
+	Validation *CustomResourceValidation `json:"validation,omitempty"`
```
@nikhita Why again is this pointer type necessar? I think you explained it before.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -136,3 +141,105 @@ type CustomResourceDefinitionList struct {
+type CustomResourceValidation struct {
+	// JSONSchema is the JSON Schema to be validated against.
+	// Can add other validation methods later if needed.
+	JSONSchema *JSONSchemaProps `json:"jsonSchema,omitempty"`
```
no need for the json tags here in the internal types

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -136,3 +141,105 @@ type CustomResourceDefinitionList struct {
+	Pattern          string        `json:"pattern,omitempty"`
+	MaxItems         *int64        `json:"maxItems,omitempty"`
+	MinItems         *int64        `json:"minItems,omitempty"`
+	// disable uniqueItems for now because it can cause the validation runtime
```
if we want that we have to move it to some validation.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -20,6 +20,9 @@ import (
+	"github.com/go-openapi/spec"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/validate"
 	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
```
nit: empty line above kube imports

sttts replied:
I would move the validation logic into `k8s.io/apiextensions-apiserver/pkg/apiserver/validation`, to keep the actual API types separate from the "business logic".

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +161,184 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+}
+
+func convertToOpenAPITypes(in *apiextensions.CustomResourceDefinition, out *spec.Schema) error {
+	if in.Spec.Validation != nil {
```
@sttts re: pointer type for CustomResourceValidation

since it is optional and would need a check here to see if it exists.

On further thought, maybe a check for `in.Spec.Validation.JSONSchema` should suffice.

sttts replied:
If our JSON codecs understand `omitempty` for non-pointer types with null-value.... If not, we need the pointer.

nikhita replied:
Will check this once more.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -40,14 +43,15 @@ type CustomResourceDefinitionStorageStrategy struct {
 }
 
-func NewStrategy(typer runtime.ObjectTyper, namespaceScoped bool, kind schema.GroupVersionKind) CustomResourceDefinitionStorageStrategy {
+func NewStrategy(typer runtime.ObjectTyper, namespaceScoped bool, kind schema.GroupVersionKind, crd apiextensions.CustomResourceDefinition) CustomResourceDefinitionStorageStrategy {
```
better use a pointer for `crd`. With nested structures ownership is not obvious anyway. We won't modify `crd` either, will we? So no need for a deepcopy either.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -113,6 +117,7 @@ func (a CustomResourceDefinitionStorageStrategy) MatchCustomResourceDefinitionSt
 type customResourceValidator struct {
 	namespaceScoped bool
 	kind            schema.GroupVersionKind
+	crd             apiextensions.CustomResourceDefinition
```
Same here, better use a pointer.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -131,6 +136,13 @@ func (a customResourceValidator) Validate(ctx genericapirequest.Context, obj run
 		return field.ErrorList{field.Invalid(field.NewPath("apiVersion"), typeAccessor.GetKind(), fmt.Sprintf("must be %v", a.kind.Group+"/"+a.kind.Version))}
 	}
 
+	customresource := obj.(*unstructured.Unstructured)
```
can we be sure that it's unstructured? If yes, let's change the signature.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +161,184 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 	return allErrs
 }
+
+func ValidateCustomResource(customresource interface{}, crd apiextensions.CustomResourceDefinition) error {
```
pointer.

Actually, there is no need to convert to OpenAPI in every call. Can we convert once and store the result in customResourceValidator?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -79,3 +80,166 @@ func TestForProperValidationErrors(t *testing.T) {
+			continue
+		}
+	}
+}
```
Can we have a test which updates a CRD validation and the checks that the new JSON schema is used?

nikhita replied:
Do you mean that after updating the CRD, the new JSONSchema is used to validate the CRs again?

sttts replied:
Had in mind this:
- create CRD
- create CR for that CRD which is rejected
- modify CRD with a less strict schema
- create CR again and see it succeeds.

Or maybe even the other way around that it gets more strict.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/helpers.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// "Constructor", parses the given string JSON reference
+func (r *JSONSchemaRef) parseJSONSchemaRef(jsonReferenceString string) error {
```
wondering which of the 5 bools in  `JSONSchemaRef` we actually need? Looking at the error handling below, we actually only parse a URL and then store that in a clever way. But actually we could should squash all this down to a simple `url.Parse` call. This decomposition is not used by us, is it?

nikhita commented:
Will rebase and squash after another review.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -350,6 +364,46 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
+		return
+	}
+
+	c.queue.Add(key)
```
I don't think we need the rate limited queue here. Our handler does one very cheap work (invalidating the cache). Just do the invalidation directly in the even handler.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -96,8 +103,15 @@ func NewCustomResourceDefinitionHandler(
 	}
 
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ret.addCustomResourceDefinition,
```
does this support to pass nil for add? we don't need add, do we?

nikhita replied:
Right, we don't need this.

_Sent from my Xiaomi Redmi Note 3 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -350,6 +364,46 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
+
+	c.queue.Add(key)
+}
+
```
don't we miss the actual invalidation on update? On seeing the c.queue.Add above.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -136,10 +139,19 @@ func (a customResourceValidator) Validate(ctx genericapirequest.Context, obj run
 
-	if err = apiextensionsvalidation.ValidateCustomResource(spec, a.crd); err != nil {
+	// Store the converted schema in CustomResourceValidator
+	if err := apiservervalidation.ConvertToOpenAPITypes(a.crd, a.schema); err != nil {
```
this conversion of the schema is supposed to happen in `NewStrategy`, isn't it?

mbohlool commented:
Sorry I didn't spend enough time to understand all of these (I will) but just as a suggestion/question, are we going to add Custom resources spec dynamically to main OpenAPI spec? (I think we should).

sttts commented:
> are we going to add Custom resources spec dynamically to main OpenAPI spec? (I think we should).

That's the plan. We have a converter from our api types into go-openapi. So it should be easy (mostly plumbing).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -40,14 +44,16 @@ type CustomResourceDefinitionStorageStrategy struct {
 			namespaceScoped: namespaceScoped,
 			kind:            kind,
+			crd:             crd,
+			schema:          &openapispec.Schema{},
```
Here I would expect the conversion from our spec into the openapi types.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -131,6 +139,22 @@ func (a customResourceValidator) Validate(ctx genericapirequest.Context, obj run
+	if !ok {
+		return field.ErrorList{field.Invalid(field.NewPath(""), customresource, fmt.Sprintf("has type %T. Must be a pointer to an Unstructured type", customresource))}
+	}
+	spec := customresource.UnstructuredContent()["spec"]
```
Why is only the spec checked?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go:
```diff
@@ -141,6 +141,78 @@ func NewCurletInstance(namespace, name string) *unstructured.Unstructured {
+	}
+}
+
+func NewNoxuValidationInstance(namespace, name string) *unstructured.Unstructured {
```
can we move this into validation_test.go?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go:
```diff
@@ -141,6 +141,78 @@ func NewCurletInstance(namespace, name string) *unstructured.Unstructured {
 	}
 }
 
+func NewNoxuValidationCRD(scope apiextensionsv1beta1.ResourceScope) *apiextensionsv1beta1.CustomResourceDefinition {
```
also this one: let's move this into validation_test.go. we won't share it with other tests.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func convertJSONSchemaProps(in *apiextensions.JSONSchemaProps, out *spec.Schema) error {
+	if in != nil {
```
can we `return nil` early when `in == nil` holds? Less indention is usually better, or at least more idiomatic in golang.

farcaller commented:
Is validation of "this field cannot be modified" part of the overall json schema plan?

sttts commented:
> Is validation of "this field cannot be modified" part of the overall json schema plan?

I like the idea to express that in a schema. But this is something on-top we could implement in a follow-up if we find a good way to express this in JSON schema (http://json-schema.org).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -348,6 +359,11 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
 
+func (c *crdHandler) updateCustomResourceDefinition(obj, _ interface{}) {
+	castObj := obj.(*apiextensions.CustomResourceDefinition)
+	glog.V(4).Infof("Updating customresourcedefinition %s", castObj.Name)
```
here the storage cache invalidation will go, right?

nikhita replied:
Done. PTAL.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go:
```diff
@@ -41,6 +41,9 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				obj.Names.ListKind = obj.Names.Kind + "List"
 			}
+			if obj.Validation.JSONSchema.Default == nil {
+				obj.Validation.JSONSchema.Default = "random-value"
```
would use a random string, `c.String()` or something like that

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/helpers.go:
```diff
@@ -0,0 +1,24 @@
+
+package v1beta1
+
+func Float64Ptr(f float64) *float64 {
```
don't we have similar helpers somewhere else already?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -348,6 +359,19 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
+
+	storageMap := c.customStorage.Load().(crdStorageMap)
+	delete(storageMap, oldCRD.UID)
+	_ = c.getServingInfoFor(newCRD)
```
I fear we have a data race now: `getServingInfoFor` reads a value from the map, here we delete it. On the other hand, we don't want to move the `Lock()` up in `getServingInfoFor`. I think we need a RWMutex, or we make a copy of the map here, change it and write the map back.

sttts replied:
The later avoid the lock for the normal code path.

sttts commented:
/ok-to-test

nikhita commented:
/retest

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -18,6 +18,11 @@ package apiextensions
+// TODO: This does not generate to protobuf because of the interface{} - fix protobuf
+//   generation to support a meta type that can accept any valid JSON.
+
+// +protobuf=false
```
these are the internal types. don't think we need this here. Only for the external ones.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -139,3 +145,86 @@ type CustomResourceDefinitionList struct {
+
+const (
+	// JSONSchemaDraft4URL is the url for JSON Schema Specification Draft 4.
+	JSONSchemaDraft4URL JSONSchemaURL = "http://json-schema.org/draft-04/schema#"
```
is this mandatory to set in `Schema`? Do we validate that? We should, shouldn't we? Just in case we support Draft4 + 1 some day.

nikhita replied:
The current behavior is that `$schema` will default to this value (but can be specified by the user).

However, since we support only draft4 now, should we not allow the user to specify this value?

sttts replied:
In any case, he should be able to specify it. My question is more whether we should require it instead of defaulting. Do we have anything in our API change guidelines about changing defaults later in a stable API? Will be able to move the default to Draft4 + 1 ever if we started with Draft4 now?

nikhita replied:
According to the guidelines, the default values should not be changed for compatibility. https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility.

nikhita replied:
/cc @deads2k do we want to maintain consistency by not using defaulting (with a good error message specifying which value to use) or do we want to make it easy for the user by using defaulting?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -139,3 +145,86 @@ type CustomResourceDefinitionList struct {
+
+// StringOrArray represents a value that can either be a string or an array of strings.
+// Mainly here for serialization purposes.
+type StringOrArray []string
```
Does this work for serialization? I mean don't we need a custom un/marshaler for that to support a single string?

nikhita replied:
For a single string, can we not use `[]string{"integer"}`?

sttts replied:
yes, we can and should. But, I wonder whether `"type": "integer"` is parsed by the json codec into `[]string{"integer"}`. Can you check (in a small test)?

nikhita replied:
`"type": "integer"` does not work. Do we need a custom marshaller/unmarshaller or `[]string{"integer"}` is fine?

sttts replied:
Is this solved? `{"type": "number"}` and `{"type": ["number", "string"]}` must de- and encode fine. A unit tests would be nice. Also make sure that the roundtrip test fuzzers cover these two cases.

nikhita replied:
Not solved. If we want this, we can have `type` as just a `string` and handle the conversion in `convertToOpenAPI`. Will add tests about this.

sttts replied:
I think we have no choice there and must implement it, because we want to decode every JSON schema, but reject our defined openapiv3 subset via validation.

sttts replied:
Compare vendor/k8s.io/apimachinery/pkg/util/intstr/intstr.go:43. That's very similar.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +159,82 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+	}
+
+	if JSONSchema.UniqueItems == true {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("JSONSchema"), "uniqueItems cannot be set to true since the runtime complexity becomes quadratic"))
```
We need a recursive walker for this because it can occur everywhere deep in the schema. Also for those below of course.

sttts replied:
taking everything back :) already implemented.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +159,82 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+
+	if len(JSONSchema.AnyOf) != 0 {
+		for _, jsonSchema := range JSONSchema.AnyOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("JSONSchema"))...)
```
the fldPath must be customized for all these recursive cases.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -348,6 +358,19 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
+	delete(storageMap, oldCRD.UID)
+	c.customStorageLock.Unlock()
+
+	_ = c.getServingInfoFor(newCRD)
```
I wonder whether we should better drop this line. It's called in the handler anyway the first time it's needed. No need to prepopulate the map.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	out.ID = in.ID
+	if in.Schema == "" {
```
Just realized this should not be here but in `apiextensions/validation`.

sttts replied:
compare `defaults.go` in other api groups.

sttts replied:
Is it checked in validation though to be equal to `apiextensions.JSONSchemaDraft4URL` ? not that somebody writes `foo bar` into it.

nikhita replied:
It's not, but will add it.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,198 @@
+)
+
+// ValidateCustomResource validates the Custom Resource against the schema in the CustomResourceDefinition.
+func ValidateCustomResource(customresource interface{}, validator *validate.SchemaValidator) error {
```
nit: add to the comment that customresource is actually a json data structure.
nit: customResource

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +159,82 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 }
+
+// ValidateCustomResourceDefinitionSchema statically validates
+func ValidateCustomResourceDefinitionSchema(JSONSchema *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
```
Can we try to convert the ref to an openapi ref here as well for validation? Then we can be sure that the actual conversion later won't error out.

nikhita replied:
Done.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -40,14 +48,26 @@ type CustomResourceDefinitionStorageStrategy struct {
+func NewStrategy(typer runtime.ObjectTyper, namespaceScoped bool, kind schema.GroupVersionKind, crd *apiextensions.CustomResourceDefinition) CustomResourceDefinitionStorageStrategy {
+	schema := &spec.Schema{}
+	if err := apiservervalidation.ConvertToOpenAPITypes(crd, schema); err != nil {
+		utilruntime.HandleError(err)
```
why not pass the openapi scheme here as parameter and do the conversion in the caller (getServingInfoFor?). There we can handle the error better than here.

nikhita replied:
That would work too. But I'm not sure how the error handling would be better there? Or is it because we can output the error earlier?

sttts replied:
we can return an error there. Here, we can't because it is a constructor. We could add an error as return value, but that would be unidiomatic. I think moving these 7 lines (or even 8 lines, including the `validate.NewSchemaValidator(schema, nil, "", strfmt.Default)` into the caller would elegantly solve this IMO.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -113,6 +133,7 @@ func (a CustomResourceDefinitionStorageStrategy) MatchCustomResourceDefinitionSt
+	validator       *validate.SchemaValidator
 }
 
 func (a customResourceValidator) Validate(ctx genericapirequest.Context, obj runtime.Object) field.ErrorList {
```
is there are reason `obj` is not a `*unstructured.Unstructured` in the function signature? Do we need the type case below?

sttts replied:
Oh, there is a reason: this is part of the Strategy interface, isn't it? So we should be pretty sure the type cast below won't error out.

nikhita replied:
Yes :)

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go:
```diff
@@ -212,7 +212,7 @@ func checkForWatchCachePrimed(crd *apiextensionsv1beta1.CustomResourceDefinition
 	}
 
-	instanceName := "foo"
+	instanceName := "setup-instance"
```
thanks :)

sttts commented on staging/src/k8s.io/apimachinery/pkg/api/testing/roundtrip/roundtrip.go:
```diff
@@ -216,10 +215,12 @@ func roundTripToAllExternalVersions(t *testing.T, scheme *runtime.Scheme, codecF
 		}
 		t.Logf("\tround tripping to %v %v", externalGVK, externalGoType)
 
+		t.Logf("\t\tjson")
```
these output line were just for  debugging. I think you can drop them again (at least here. I would like to see an independent PR add them to the existing t.Logf's, i.e. to avoid extra lines in the test output).

sttts commented on staging/src/k8s.io/kube-gen/cmd/conversion-gen/generators/conversion.go:
```diff
@@ -328,7 +328,7 @@ func (e equalMemoryTypes) Skip(a, b *types.Type) {
 }
 
-func (e equalMemoryTypes) Equal(a, b *types.Type) bool {
+func (e equalMemoryTypes) Equal(a, b *types.Type, currentlyBuildingTypes map[*types.Type]bool) bool {
```
Can you prepare an independent PR for this?

nikhita replied:
Will do!

nikhita replied:
Created https://github.com/kubernetes/kubernetes/pull/49747.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,198 @@
+	out.Pattern = in.Pattern
+	out.MaxItems = in.MaxItems
+	out.MinItems = in.MinItems
+	// disable uniqueItems because it can cause the validation runtime
```
this shouldn't be necessary if we have the validation.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,198 @@
+		return err
+	}
+
+	// need a check when the Ref is empty because NewRef gives a "#", instead of an empty ref
```
can you link the issue here?

nikhita replied:
Done.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +161,92 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+
+	if len(JSONSchema.Properties) != 0 {
+		for property, jsonSchema := range JSONSchema.Properties {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("Properties."+property))...)
```
PTAL this please. :)

The fldPath will now be something like `spec.Validation.JSONSchema.Properties.zeta.UniqueItems`. I am not sure if it is a good idea to have the `zeta` part? But it denotes _which property_ the child fields are under.

The other alternative is to have the fldPath like this: `spec.Validation.JSONSchema.Properties.UniqueItems`. This does not tell us the property under which `UniqueItems` lies.

nikhita commented:
The tests won't pass since the conversion-gen change no longer lies in this PR. They should pass once #49747 is merged.

nikhita commented:
/test pull-kubernetes-unit

nikhita commented:
/joke

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/validation.go:
```diff
@@ -0,0 +1,160 @@
+	}
+
+	if in.Ref != nil {
+		out.Ref, err = spec.NewRef(*in.Ref)
```
:+1:

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -387,9 +387,17 @@ func (c *crdHandler) updateCustomResourceDefinition(oldObj, _ interface{}) {
 	storageMap := c.customStorage.Load().(crdStorageMap)
-	delete(storageMap, oldCRD.UID)
-	c.customStorageLock.Unlock()
+	storageMap2 := make(crdStorageMap)
```
`make(crdStorageMap, len(storageMap))` saves some allocations

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -387,9 +387,17 @@ func (c *crdHandler) updateCustomResourceDefinition(oldObj, _ interface{}) {
-	c.customStorageLock.Unlock()
+	storageMap2 := make(crdStorageMap)
+
+	for k, v := range storageMap {
```
can you put a comment above like "Copy because we cannot write to storageMap without a race as it is used without locking elsewhere

sttts replied:
fixed

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -387,9 +387,17 @@ func (c *crdHandler) updateCustomResourceDefinition(oldObj, _ interface{}) {
+	}
+
+	delete(storageMap2, oldCRD.UID)
+	c.customStorage.Store(storageMap2)
```
We need the same for adding an element.

nikhita replied:
Created https://github.com/kubernetes/kubernetes/pull/50098.

nikhita commented:
@deads2k Will address the discussion from https://github.com/kubernetes/kubernetes/issues/49879. But could you take a look at this in general and see if this looks good to you? :)

nikhita commented:
Updated to support OpenAPISpecV2 and OpenAPISpecV3. Specifically, this commit: https://github.com/kubernetes/kubernetes/pull/47263/commits/74e606c54a839cdc8b8c90b895832163874657e5.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -164,11 +164,12 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 
-type openAPISpec apiextensions.JSONSchemaProps
+type additionalSpecValidatorV2 struct{}
+
```
nit: move them down where the methods are

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,144 +162,173 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 
-	allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISpec(OpenAPISpecV2, fldPath.Child(""))...)
+	openAPISpecV3 := &additionalSpecValidatorV3{}
+	allErrs = append(allErrs, openAPISpecV3.validateCustomResourceDefinitionOpenAPISpec(schema, fldPath)...)
```
maybe a comment that all of v3 also applies to v2.

nikhita replied:
Done.

mbohlool replied:
If the checks applied to both v2 and v3, why don't we do them in the common validator function above? In theory, the additional validator for v3 should be empty.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -139,3 +140,91 @@ type CustomResourceDefinitionList struct {
+// Can add other validation methods later if needed.
+type CustomResourceValidation struct {
+	// OpenAPISpecV2 is the OpenAPI v2 schema to be validated against.
+	OpenAPISpecV2 *JSONSchemaProps
```
As @enisoc suggested (https://github.com/kubernetes/kubernetes/issues/49879#issuecomment-321407614), OpenAPIV2Schema is a better name.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +161,176 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+
+	if len(schema.Type) != 0 {
+		if len(schema.Type) > 1 {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("type"), "multiple values via an array for Type is not supported."))
```
for type

sttts commented:
@enisoc @deads2k @mbohlool please take a look here. From my side it's lgtm @enisoc's suggested renaming to `OpenAPIV{2,3}Schema`, but let's have some more eyes on it.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,144 +162,174 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 
+// additionalSpecValidator applies validations for different OpenAPI specfication versions.
+type additionalSpecValidator interface {
+	validateCustomResourceDefinitionOpenAPISpec(spec *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList
```
I would simply call this function `validate`. The interface (and the type name) are descriptive enough about what it is validating.

sttts replied:
Validate is better, I agree.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,144 +162,174 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+
 // ValidateCustomResourceDefinitionValidation statically validates
-func ValidateCustomResourceDefinitionValidation(CustomResourceValidation *apiextensions.CustomResourceValidation, fldPath *field.Path) field.ErrorList {
+func ValidateCustomResourceDefinitionValidation(customResourceValidation *apiextensions.CustomResourceValidation, fldPath *field.Path) field.ErrorList {
```
The name has two `Validate` (in different forms) in it. Would it be better to name this `ValidateCustomResourceDefinitionSpec`?

nikhita replied:
Yeah, the name looks a bit funny. But `ValidateCustomResourceDefinitionSpec` exists already :/

(https://github.com/kubernetes/kubernetes/pull/47263/files#diff-54fdd67617b0f2ef1b5a09148f77303eR64-R107).

sttts replied:
I would keep the name. It's logical. Better something logical and consistent than something shorter/different, but inconsistent.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,144 +162,174 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
-// ValidateCustomResourceDefinitionOpenAPISpecV2 statically validates
-func ValidateCustomResourceDefinitionOpenAPISpecV2(OpenAPISpecV2 *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
+// ValidateCustomResourceDefinitionOpenAPISpec statically validates
+func ValidateCustomResourceDefinitionOpenAPISpec(schema *apiextensions.JSONSchemaProps, fldPath *field.Path, specVersion additionalSpecValidator) field.ErrorList {
```
`specVersion` is a misleading name. Would it better to be called `versionSpecificValidator` or simply `additionalValidator`?

sttts replied:
I would rename 
- `additionalSpecValidator` to `specStandardValidator`
- `specVersion` to `ssv` or `specStandardValidator`

In addition, I would move all non-datatype validations (i.e. any kind of restrictions that do not apply to generic JSON Schema) into the v3 specStandardValidator.

sttts replied:
I.e. checking `$ref` in the common part is fine (that's about the datatype). But forbidding array `type` is v3.

mbohlool commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,144 +162,174 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
-		if len(OpenAPISpec.Items.JSONSchemas) != 0 {
-			allErrs = append(allErrs, field.Forbidden(fldPath.Child("items"), "items must be a schema object and not an array"))
-		}
+	if schema.Items != nil && len(schema.Items.JSONSchemas) != 0 {
```
Duplicate of the check in common validator?

mbohlool commented:
Mostly name suggestion. Would be nice if you can squash commits and make them logically separated. They are representing stages of reviews now. But overall it looks good.

nikhita committed:
Update godeps

nikhita commented:
Addressed comments. Removed v2. Rebased and squashed.

nikhita commented:
/retest

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -139,3 +140,89 @@ type CustomResourceDefinitionList struct {
 }
+
+// CustomResourceValidation is a list of validation methods for CustomResources.
+// Can add other validation methods later if needed.
```
Would remove `Can add other validation methods later if needed.`. Also in the versioned type.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -26,9 +26,10 @@ type CustomResourceDefinitionSpec struct {
 	// Scope indicates whether this resource is cluster or namespace scoped.  Default is namespaced
 	Scope ResourceScope
+	// Validation describes the validation methods for CustomResources
+	Validation CustomResourceValidation
```
do we have a convention for mark alpha fields? There was a PR out from @liggitt to update the docs for that.

nikhita replied:
Added.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/helpers.go:
```diff
@@ -109,3 +109,11 @@ func CRDRemoveFinalizer(crd *CustomResourceDefinition, needle string) {
 	crd.Finalizers = newFinalizers
 }
+
+func Float64Ptr(f float64) *float64 {
```
Are these only used in validation_test.go? Can we move them there and make them private?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +161,92 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+	}
+
+	if JSONSchema.AdditionalItems != nil {
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(JSONSchema.AdditionalItems.Schema, fldPath.Child("AdditionalItems"))...)
```
all the `fldPath.Child(...)` should be lower-case, i.e. the json field names.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -287,6 +308,19 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
+	if err := apiservervalidation.ConvertToOpenAPITypes(crd, openapiSchema); err != nil {
+		return nil, err
+	}
+
```
nit: no empty lines here and below

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -131,6 +137,16 @@ func (a customResourceValidator) Validate(ctx genericapirequest.Context, obj run
 
+	customResourceObject, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return field.ErrorList{field.Invalid(field.NewPath(""), customResourceObject, fmt.Sprintf("has type %T. Must be a pointer to an Unstructured type", customResourceObject))}
```
this is actually a fatal programming error. But I guess we have no way to return that here properly. So let's leave it under the label: this will never happen.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -101,7 +101,7 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 
 	allErrs = append(allErrs, ValidateCustomResourceDefinitionNames(&spec.Names, fldPath.Child("names"))...)
-	allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(spec.Validation.JSONSchema, fldPath.Child("Validation.JSONSchema"))...)
+	allErrs = append(allErrs, ValidateCustomResourceDefinitionValidation(&spec.Validation, fldPath.Child("validation"))...)
```
this commit "Validate CRD Schema" can be squashed into "Validate CustomResource  …", can't it?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,91 +162,161 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+
+	if customResourceValidation.OpenAPIV3Schema != nil {
+		openAPIV3Schema := &specStandardValidatorV3{}
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(customResourceValidation.OpenAPIV3Schema, fldPath.Child("openAPIV3Schema"), openAPIV3Schema)...)
```
no `.Child` here. It's the same object, only additional checks.

nikhita replied:
This would make the JSON path as `validation.openAPIV3Schema.type`. Without the `.Child`, we get `validation.type`. Do we want the latter?

sttts replied:
you are right. I thought we are inside the `openAPIV3Schema` here already.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -162,91 +162,161 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("Properties."+property))...)
+	if len(schema.Properties) != 0 {
+		for property, jsonSchema := range schema.Properties {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("properties["+property+"]"), ssv)...)
```
`"properties["+property+"]"` => `fmt.Sprintf("properties[%q]", property)`

sttts replied:
q for quotes (is that correct?)

nikhita replied:
q is for double-quoted string safely escaped with Go syntax. Eg: `"\"string\""`

sttts replied:
my question was: how does JSON path encode this? With or without quotes?

nikhita replied:
we have this https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/field/path.go#L64 :)

nikhita committed:
Add feature gate for CustomResourceValidation

update feature gates for generic apiserver

Add apiextensions-apiserver features to golint_failures

Ignore alpha feature if gate is disabled

sttts commented on pkg/features/kube_features.go:
```diff
@@ -133,6 +133,12 @@ const (
+	// alpha: v1.8
+	//
+	// CustomResourceValidation is a list of validation methods for CustomResources
+	CustomResourceValidation utilfeature.Feature = "CustomResourceValidation"
```
I think we can reference the one in `staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go` here, can't we? Or even only below and drop this paragraph compeltely. Just follow `AdvancedAuditing`.

nikhita commented:
verify fails with:

```
W0815 09:18:28.018] staging/src/k8s.io/apiextensions-apiserver/pkg/features/kube_features.go:30:2: comment on exported const CustomResourceValidation should be of the form "CustomResourceValidation ..."
W0815 09:18:28.018] 
W0815 09:18:28.018] Please review the above warnings. You can test via "golint" and commit the result.
W0815 09:18:28.019] If the above warnings do not make sense, you can exempt this package from golint
W0815 09:18:28.019] checking by adding it to hack/.golint_failures (if your reviewer is okay with it).
W0815 09:18:28.019] 
I0815 09:18:28.119] FAILED   hack/make-rules/../../hack/verify-golint.sh	61s
```

Added `staging/src/k8s.io/apiextensions-apiserver/pkg/features` to `./hack/.golint_failures`. Other feature packages are also present there.

sttts commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -47,6 +47,13 @@ const (
+	//
+	// Taint nodes based on their condition status for 'NetworkUnavailable',
+	// 'MemoryPressure', 'OutOfDisk' and 'DiskPressure'.
+	TaintNodesByCondition utilfeature.Feature = "TaintNodesByCondition"
```
Why was this moved?

nikhita replied:
This is present in the generic api-server: https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L166.

sttts replied:
Looks like it's in master, not the generic apiserver. Then reference the master feature here.

sttts commented on pkg/features/kube_features.go:
```diff
@@ -161,7 +148,12 @@ var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureS
+	genericfeatures.StreamingProxyRedirects: {Default: true, PreRelease: utilfeature.Beta},
+	genericfeatures.AdvancedAuditing:        {Default: false, PreRelease: utilfeature.Alpha},
+	genericfeatures.APIResponseCompression:  {Default: false, PreRelease: utilfeature.Alpha},
+	genericfeatures.TaintNodesByCondition:   {Default: false, PreRelease: utilfeature.Alpha},
```
I think TaintNodesByCondition should be defined in this file (nodes are a kube concept, not generic). It was only put under the wrong heading, i.e. should be in the list above.

nikhita replied:
Will update.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -100,6 +104,12 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceValidation) {
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionValidation(&spec.Validation, fldPath.Child("validation"))...)
+	} else if spec.Validation.OpenAPIV3Schema != nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("validation"), "disabled by feature-gate"))
```
what does the alpha field policy say? Should we just set this to zero on create and update or should we reject it?

nikhita replied:
set it to zero. From the policy: `ensure the field is entirely absent from the API when empty (might require making the field a pointer)`.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -139,3 +147,88 @@ type CustomResourceDefinitionList struct {
+	Schema               JSONSchemaURL              `json:"$schema,omitempty"`
+	Ref                  *string                    `json:"$ref,omitempty"`
+	Description          string                     `json:"description,omitempty"`
+	Type                 StringOrArray              `json:"type,omitempty"`
```
Does StringArrayOrString work here?

nikhita replied:
Works perfectly. :)

sttts commented on pkg/features/kube_features.go:
```diff
@@ -158,10 +152,15 @@ var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureS
 	DebugContainers:                             {Default: false, PreRelease: utilfeature.Alpha},
 	PodPriority:                                 {Default: false, PreRelease: utilfeature.Alpha},
 	EnableEquivalenceClassCache:                 {Default: false, PreRelease: utilfeature.Alpha},
+	TaintNodesByCondition:                       {Default: false, PreRelease: utilfeature.Alpha},
```
or reference the pkg/master.TaintNodesByCondition

nikhita replied:
From the offline discussion: looks like there is no TaintNodesByCondition in pkg/master so leaving this as-is.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -19,12 +19,25 @@ package integration
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
+func float64Ptr(f float64) *float64 {
```
nit: usually those funcs go to the bottom

nikhita replied:
Done.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +167,166 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+	allErrs = append(allErrs, ssv.validate(schema, fldPath)...)
+
+	if schema.UniqueItems == true {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("uniqueItems"), "uniqueItems cannot be set to true since the runtime complexity becomes quadratic"))
```
can you check that the error message include the field name usually? I thought the output would be something like `fldPath + " " + err`

sttts replied:
compare e.g. https://github.com/kubernetes/kubernetes/pull/47263/commits/cbab6d02ec026d6c7808a9d89a41e79bcf84b818#diff-34aedeefa7784abbe7531bae68833ba6R144

nikhita replied:
the error message is: `CustomResourceDefinition.apiextensions.k8s.io "noxus.mygroup.example.com" is invalid: spec.validation.openAPIV3Schema.additionalProperties: Forbidden: additionalProperties cannot be set to false` which looks fine. So, leaving this as-is.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -158,3 +167,166 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
+
+	if schema.AdditionalProperties != nil {
+		if schema.AdditionalProperties.Allows == false {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("additionalProperties"), "additionalProperties cannot be set to false"))
```
maybe a comment above: additionalProperties contradict Kubernetes API convention to ignore unknown fields

nikhita replied:
Done.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -100,6 +105,10 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 
 	allErrs = append(allErrs, ValidateCustomResourceDefinitionNames(&spec.Names, fldPath.Child("names"))...)
 
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceValidation) {
```
if the flag is off, you should disallow setting it: https://github.com/kubernetes/community/pull/869

nikhita replied:
@deads2k Had done that earlier. But after this discussion: https://github.com/kubernetes/community/pull/913#discussion_r133227732, it was decided to drop the feature: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-18bbf71961e0d43572fc7357183260feR51

liggitt replied:
belt and suspenders... I would clear it (but not in validation... maybe in PrepareForCreate/PrepareForUpdate?) *and* ensure it is not set in validation

sttts replied:
That's done as well.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -98,6 +107,10 @@ func NewCustomResourceDefinitionHandler(
 	}
 
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: ret.updateCustomResourceDefinition,
```
It's unusual to handle update and not create.  Why the difference?

sttts replied:
create is implict. We create the strategy on demand (your code btw ;-)).

But on update we have to delete the cache strategy. That's what this handler does.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -287,6 +308,17 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
 	creator := unstructuredCreator{}
+
+	// convert CRD schema to openapi schema
+	openapiSchema := &openapispec.Schema{}
```
Is this block likely to fail?  If so, this reads like a status condition

sttts replied:
No, the validation should make sure it does not fail. Everything else is a bug in validation.

sttts replied:
No, the validation should make sure it does not fail. Everything else is a bug in validation.

sttts replied:
Also note that the conversion is basically a 1:1 translation with a number of restrictions on the input, plus some string->openapi-type conversion which are done during validation already in order to guarantee they work in this step as well.

deads2k commented:
looks good at a high level.

deads2k commented:
/approve

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go:
```diff
@@ -72,6 +72,18 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				obj.Example = "foobarbaz"
 			}
+			if c.RandBool() {
+				c.Fuzz(obj.Items)
```
which type are items? Aren't they `interface{}`?

sttts replied:
oh, array or prop, right?

nikhita replied:
`Items` is `*JSONSchemaPropsOrArray`

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go:
```diff
@@ -67,11 +67,43 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				obj.Example = "foobarbaz"
 			}
 		},
+		func(obj *apiextensions.JSONSchemaPropsOrBool, c fuzz.Continue) {
```
:+1:

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go:
```diff
@@ -67,11 +67,43 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+			}
+		},
+		func(obj *apiextensions.JSONSchemaPropsOrArray, c fuzz.Continue) {
+			// disallow both Schema and JSONSchemas to be nil.
```
and not both should be non-nil at the same time. I would use a RandBool if-clause as above.

nikhita replied:
Done.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/fuzzer/fuzzer.go:
```diff
@@ -67,11 +67,43 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+			c.Fuzz(obj.Schema)
+			if obj.Schema == nil {
+				c.Fuzz(&obj.JSONSchemas)
+				if len(obj.JSONSchemas) == 0 {
```
you can also create the array yourself and then call c.Fuzz on each element. Would maybe be slightly cleaner.

sttts commented:
/lgtm

smarterclayton commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -139,3 +140,84 @@ type CustomResourceDefinitionList struct {
+// CustomResourceValidation is a list of validation methods for CustomResources.
+type CustomResourceValidation struct {
+	// OpenAPIV3Schema is the OpenAPI v3 schema to be validated against.
+	OpenAPIV3Schema *JSONSchemaProps
```
How big are these normally?  What max size limit is imposed (our 1m value limit)?

sttts replied:
We cannot rely on an external library to basically do our validation and to follow our api guidelines. So yes, it's about compatibility. We were discussing this in the proposal up and down and voted for this solution for essentially the reason of having control. Still I don't get the argument. The complexity is not big, look at the validations. It's completely straight forward. What is left is a proto marshaller.

mbohlool replied:
What is discussed before was possibility of including the external type in our API (that I agree not a good idea) but the suggestion here is to use rawExtension. What we unmarshall it to is totally an implementation detail not a design one. So two things here:

- What exposed in our API! Can be RawExtension with a comment that this is standard Vxxx. Or can be our defined type (that may be slightly different than standard)
- How we implement it internally. With RawExtension we can unmarshall it to an internal type or go-openapi type. It is not user faced. With our defined type exposed, no options here.

The reason I think rawExtension is a good idea is that we don't/can't apply our API standards to an already defined standard (without changing the standard, e.g. StringOrStringArray). Having control over spec means we can diverge from the standard and I don't think having that power is a good thing.

Having RawExtension can also let us support different version of Spec without adding new fields. The standard has the version in it, so we can just look at it first before unmarshall it to the right type.

sttts replied:
We discussed RawExtension as well. It does not make a big difference how the unmarshaling works, whether via RawExtenions or our own types. In the end we define an external API surface we have to support and RawExtensions are an implementation detail to delegate the unmarshaling. It does not solve any of the API validation and stability problems. Moreover, we know that go-openapi's validation of JSON Schema is weak at best. IMO we cannot trust it for this purpose and have no control over it. We also discussed to use the `$scheme` field for the version before and decided against it (that's why we have `OpenAPIv3` now btw).

In general, we have been through all of these considerations in the proposal and yet again for some here in the implementation PR. I would very much appreciate participation in the proposal process for these kinds of discussions. The proposal was posted to the relevant lists and people were CC'ed on the proposal PR and this implementation PR very early on. It's pretty tiresome to reopen these discussions again and again so late in the process when the PR is ready.

mbohlool replied:
You are completely right. I am sorry that we didn't catch this in the design phase. But I am just worried that we add something here that we would stuck supporting while there is, to my opinion, a better option. Using RawExtension is not an implementation detail here and should not impose a lot of changes (we can still keep our type just make it internal, if we want to use go-openapi or not is at another level that can be changed or kept later without changing the surface of our API). I would leave this to @lavalamp to give his final opinion.

deads2k replied:
> If we don't actually use the schema and pass it through to other libraries, then do we really want to bless it by writing it up in our API language? The only reason I see to do that is to ensure backwards compatibility, and I'm not sure if that's persuasive.

The compatibility and control over the serialization both matter since a change in a backing API can easily produce breaking API changes in your exposed (and already serialized!) objects.  In OpenShift, we found this out the hard way by "hiding" a docker library behind a raw serialization.  We could still deserialize the code, but it not longer worked.  If we want it in our API, with our stability guarantees, saying "the backing library changed so that json you sent us broke, it wasn't *really* part of our API" isn't a stance that is consistent with the API guarantees that we have.

We must control the serialization.

smarterclayton commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -18,6 +18,11 @@ package v1beta1
+// TODO: This does not generate to protobuf because of the interface{} - fix protobuf
+// generation to support a meta type that can accept any valid JSON.
+
+// +protobuf=false
```
This is a breaking API change, to disable protobuf.

sttts replied:
@lavalamp don't worry. We won't. Solution is on its way.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go:
```diff
@@ -139,3 +140,84 @@ type CustomResourceDefinitionList struct {
+// CustomResourceValidation is a list of validation methods for CustomResources.
+type CustomResourceValidation struct {
+	// OpenAPIV3Schema is the OpenAPI v3 schema to be validated against.
+	OpenAPIV3Schema *JSONSchemaProps
```
Per CRD? more like 10-100 KB. It's per type.

mbohlool replied:
I think @lavalamp suggested we change this to a json string instead of representing Schema in our struct. Does that make things easier? (e.g. does problem of protobuf @smarterclayton pointed out will fade away if we do that) I think representing this in our API made lots of trouble for us as we don't really benefit from it. We can accept an string (or a free object) and deserialize it to JSONSchemaProps at our backend code. WDYT?

sttts replied:
It doesn't make it easier.

sttts replied:
The only left issue is protobuf. And we have two working approaches. Just a little patience.

mbohlool replied:
I know we can make it work and it is hard to throw away code/reviews. But other than cost of coding, there would be const of maintenance and bugs etc. Adding it as RawExtension would only add a unmarshall call to this code and get rid of many types/conversions/issues (solved and unsolved). I personally vote for simpler code in this case. We don't need to know the content of validation schema and we won't change it (unless we change the whole version of it). That is an standard spec.

lavalamp replied:
If we don't actually use the schema and pass it through to other libraries, then do we really want to bless it by writing it up in our API language? The only reason I see to do that is to ensure backwards compatibility, and I'm not sure if that's persuasive.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -18,6 +18,11 @@ package v1beta1
+// TODO: This does not generate to protobuf because of the interface{} - fix protobuf
+// generation to support a meta type that can accept any valid JSON.
+
+// +protobuf=false
```
protobuf generation does not work due to the presence of interface fields.

We can, however, have an intermediate representation for these fields and define a marshaller for proto on that.

We can have two cases for this intermediate representation:

1. A struct like:

```
type JSON struct {
   Float *float
   String *string
   Map map[string]JSON
   ...
}
```

2. A string.

@smarterclayton wdyt?

sttts replied:
And probably an alias for `interface{}` like `type JSON interface{}` to write the custom marshaller for. @smarterclayton had you tried that for the table type?

sttts replied:
One way would be by having @nikhita's `JSON` in the external types and `interface{}` in the internal. Then we can use normal proto tags on `JSON` and custom JSON marshal/unmarshal methods. Alternatively custom marshal/unmarshal for protobuf if that exists.

smarterclayton replied:
Yeah you just need to have a custom marshaller.  runtime.RawExtension is a *very* optimized version of this, but you can look at Time.  Basically you want to turn the struct into json, and then unmarshal a blob of bytes back to your own struct.  Should be easy.

lavalamp replied:
Definitely can't turn proto off for this entire file.

nikhita commented:
/test pull-kubernetes-e2e-gce-gpu

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -139,3 +147,84 @@ type CustomResourceDefinitionList struct {
+}
+
+// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
+type JSONSchemaProps struct {
```
Presumably this was covered in a design doc that I didn't manage to look at, but: is it really a good idea to recreate a spec type in our API like this? Was there a debate on the pros and cons? Sorry to bring this up at this late stage. But stuff we put in our API is forever.

sttts replied:
What exactly is the question? Yes, there was debate. Even if we reused types from a library, it would be in our API forever. The point is that this way we can guarantee it forever. By using a library which meanwhile switched to OpenAPI v6 we maybe couldn't.

nikhita commented:
cc @deads2k ready for review.

sttts commented:
/retest

nikhita commented:
/test pull-kubernetes-bazel-build

sttts commented:
@smarterclayton @deads2k protobuf support is added now, with a stripped down variant of `runtime.Unknown`. please take a look.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+func (s JSON) MarshalJSON() ([]byte, error) {
+	if len(s.Raw) > 0 {
```
note that you're compressing empty into "null".  That's reasonable given the rest of the API, it just catches my eye.

deads2k commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/conversion.go:
```diff
@@ -0,0 +1,73 @@
+func Convert_v1beta1_JSON_To_apiextensions_JSON(in *JSON, out *apiextensions.JSON, s conversion.Scope) error {
+	if in != nil {
+		var i interface{}
+		if err := json.Unmarshal(in.Raw, &i); err != nil {
```
This makes sense.  The end result is a protobuf shell with a delicious json nougat-y center, right?

sttts replied:
exactly :)

nikhita replied:
Yes :)

sttts commented:
As the JSON protobuf change is squashed already, here are the relevant lines of code:

- the internal JSON type (just an alias for interface{}): https://github.com/kubernetes/kubernetes/pull/47263/files#diff-48b8b5215523ba995d8b41c2dfdf68afR192
- the external type: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-3029a18192feb70af52de676209da9b6R194, 
- an (optional) JSON field: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-48b8b5215523ba995d8b41c2dfdf68afR159, 
- the json marshal/unmarshal: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-3fe7a62396a7708e034adc5c0bbff3fdR121
- the conversion: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-a48a0d878266c677cd0c5c9987b327b6R40

@smarterclayton ^^^

deads2k commented:
> As the JSON protobuf change is squashed already, here are the relevant lines of code:
> 
> the internal JSON type (just an alias for interface{}): https://github.com/kubernetes/kubernetes/pull/47263/files#diff-48b8b5215523ba995d8b41c2dfdf68afR192
> the external type: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-3029a18192feb70af52de676209da9b6R194,
> an (optional) JSON field: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-48b8b5215523ba995d8b41c2dfdf68afR159,
> the json marshal/unmarshal: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-3fe7a62396a7708e034adc5c0bbff3fdR121
> the conversion: https://github.com/kubernetes/kubernetes/pull/47263/files#diff-a48a0d878266c677cd0c5c9987b327b6R40

This looks fairly straightforward.  protobuf wrapping a limited json field with simple deserialization and a little validation.  This may actually have been what the service catalog was looking for a while back.

smarterclayton commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go:
```diff
@@ -139,3 +142,90 @@ type CustomResourceDefinitionList struct {
+}
+
+// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
+type JSONSchemaProps struct {
```
Maybe move this into its own types.go file - `types_jsonschema.go`

smarterclayton commented:
Looks like I expected.  This has high level approval from me.

sttts commented:
/lgtm

smarterclayton commented:
/approve
